dstore: bringing to mca

karasevb opened the pull request:
In progress. Will be fixed:
- [x] ./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:0-2;1:3]"
- [x] ./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:;1:]"
- [x] ./pmix_test -n 4 --ns-dist 3:1 --fence "[0:]"
- [x] ./pmix_test -n 4 --ns-dist 3:1 --fence "[b | 0:]"
- [x] ./pmix_test -n 4 --ns-dist 3:1 --fence "[d | 0:]" --noise "[0:0,1]"
- [x] ./pmix_test -n 4 --job-fence -c
- [x] ./pmix_test -n 4 --job-fence
- [x] ./pmix_test -n 2 --test-publish
- [x] ./pmix_test -n 2 --test-spawn
- [x] ./pmix_test -n 2 --test-connect
- [x]  ./pmix_test -n 5 --test-resolve-peers --ns-dist "1:2:2"
- [x] ./pmix_test -n 5 --test-replace 100:0,1,10,50,99
- [x] ./pmix_test -n 5 --test-internal 10

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

rhc54 commented:
@karasevb Just an FYI. The ```test-resolve-peers``` test is forcing a change to the GDS interface. I should finish it today. It might actually help make your components easier - just a bit of a pain for mine.

rhc54 commented:
I'm going to update the branch, which might cause some conflicts with your commits. However, I don't want to mess you up by merging this pull request before you are ready.

karasevb commented:
Updated progress status

rhc54 commented:
Just FYI: all tests appear to be working now without dstore. Many thanks for the assist! Let me know when you are ready and I'll pull your changes into my branch

karasevb committed:
Fixes the type of GDS external module which client tell to the server

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

karasevb committed:
dstore mca implementation

./check.sh
OK ./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:0-2;1:3]"
OK ./pmix_test -n 4 --ns-dist 3:1 --fence "[db | 0:;1:]"
OK ./pmix_test -n 4 --ns-dist 3:1 --fence "[0:]"
OK ./pmix_test -n 4 --ns-dist 3:1 --fence "[b | 0:]"
OK ./pmix_test -n 4 --ns-dist 3:1 --fence "[d | 0:]" --noise "[0:0,1]"
OK ./pmix_test -n 4 --job-fence -c
OK ./pmix_test -n 4 --job-fence
OK ./pmix_test -n 2 --test-publish
OK ./pmix_test -n 2 --test-spawn
OK ./pmix_test -n 2 --test-connect
FAIL ./pmix_test -n 5 --test-resolve-peers --ns-dist "1:2:2"
OK ./pmix_test -n 5 --test-replace 100:0,1,10,50,99
FAIL ./pmix_test -n 5 --test-internal 10

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

karasevb commented:
Synchronized with main PR. I'm working for a resolve peers problem.

karasevb committed:
Fixed providing a job info for dstore

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

karasevb commented:
@rhc54 all tests passed with dstore. I will do the code cleanup and additional checks, then we can merge. I'll let you know when it's ready.

rhc54 commented:
Very good - thank you!

karasevb committed:
Added the register nspace invoke for all active GDS modules.

This fix is required when there are more than one active GDS modules.
Server must prepare a nspace for each active GDS module
before client is started. Because server doesn't know what kind of
module will be used on the client side when it start. Consequently,
there must be a nspace deregistration for each GDS during the finilize.

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

rhc54 commented on src/mca/gds/gds.h:
```diff
@@ -289,10 +289,13 @@ typedef pmix_status_t (*pmix_gds_base_module_add_nspace_fn_t)(const char *nspace
+        pmix_gds_base_active_module_t *_g;                  \
+        PMIX_LIST_FOREACH(_g, &pmix_gds_globals.actives,    \
+                          pmix_gds_base_active_module_t) {  \
+            (s) = _g->module->add_nspace(n, i, ni);         \
```
Sorry about that - the comment didn't go where I put it. Anyway, I was just suggesting that you check for NULL  before calling the function

rhc54 commented on src/mca/gds/gds.h:
```diff
@@ -306,10 +309,13 @@ typedef pmix_status_t (*pmix_gds_base_module_add_nspace_fn_t)(const char *nspace
+        pmix_gds_base_active_module_t *_g;                  \
+        PMIX_LIST_FOREACH(_g, &pmix_gds_globals.actives,    \
+                          pmix_gds_base_active_module_t) {  \
+            (s) = _g->module->del_nspace(n);                \
```
and same here

karasevb commented:
@rhc54 Thanks, I added the checking for NULL

karasevb committed:
GDS external module select by "pmix.gds.mod"

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

karasevb committed:
- Fixed the dstore module debug print output like to the mca components
- Added the macro function for unloading the PMIX_INFO array to the pmix list
- Removed unneeded file

Signed-off-by: Boris Karasev <karasev.b@gmail.com>

karasevb commented:
@rhc54 ready to merge

rhc54 commented:
Very good - thanks! I'm still thinking about your email question. It seems we always need the hash module, and so we may need to move it to the "base" so it is always available. Otherwise, it could be ignored using MCA parameters.

I'll continue to think about it - we have a little time before this can come into master anyway as we don't want it committed until after the v2.0 branch has been created.
